mlfinlab/portfolio_optimization/hcaa.py

```diff
@@ -0,0 +1,421 @@
+                                                                                           covariance=covariance_matrix,
+                                                                                           confidence_level=confidence_level,
+                                                                                           cluster_indices=right_cluster)
+                    alloc_factor = \
+                        1 - left_cluster_expected_shortfall / (left_cluster_expected_shortfall + right_cluster_expected_shortfall)
+                elif allocation_metric == 'conditional_drawdown_risk':
+                    left_cluster_conditional_drawdown = self._get_cluster_conditional_drawdown_at_risk(asset_returns=asset_returns,
+                                                         covariance=covariance_matrix,
+                                                         confidence_level=confidence_level,
+                                                         cluster_indices=left_cluster)
+                    right_cluster_conditional_drawdown = self._get_cluster_conditional_drawdown_at_risk(asset_returns=asset_returns,
+                                                         covariance=covariance_matrix,
+                                                         confidence_level=confidence_level,
+                                                         cluster_indices=right_cluster)
+                    alloc_factor = \
+                        1 - left_cluster_conditional_drawdown / (left_cluster_conditional_drawdown + right_cluster_conditional_drawdown)
+                else:
+                    alloc_factor = 0.5 # equal weighting
+
+                # Assign weights to each sub-cluster
+                self.weights[left_cluster] *= alloc_factor
+                self.weights[right_cluster] *= 1 - alloc_factor
+
+        # Assign actual asset values to weight index
+        self.weights.index = assets[self.ordered_indices]
+        self.weights = pd.DataFrame(self.weights)
+        self.weights = self.weights.T
+
+    @staticmethod
+    def _cov2corr(covariance):
+        '''
+        Calculate the correlations from asset returns covariance matrix.
+
+        :param covariance: (pd.DataFrame) asset returns covariances
+        :return: (pd.DataFrame) correlations between asset returns
+        '''
+
+        d_matrix = np.zeros_like(covariance)
+        diagnoal_sqrt = np.sqrt(np.diag(covariance))
+        np.fill_diagonal(d_matrix, diagnoal_sqrt)
+        d_inv = np.linalg.inv(d_matrix)
+        corr = np.dot(np.dot(d_inv, covariance), d_inv)
+        corr = pd.DataFrame(corr, index=covariance.columns, columns=covariance.columns)
+        return corr
+
+    @staticmethod
+    def _perform_checks(asset_prices, asset_returns, covariance_matrix, allocation_metric):
+        # pylint: disable=bad-continuation
+        '''
+        Perform initial warning checks
```

Comment:
Missing full stop. Please check Grammar for all docstrings.

Reply:
Resolved